src/bijectors/ordered.jl

```diff
@@ -17,14 +25,22 @@ Return a `Distribution` whose support are ordered vectors, i.e., vectors with in
 This transformation is currently only supported for otherwise unconstrained distributions.
 """
 function ordered(d::ContinuousMultivariateDistribution)
-    if bijector(d) !== identity
-        throw(
-            ArgumentError(
-                "ordered transform is currently only supported for unconstrained distributions.",
-            ),
-        )
+    # We're good if the map from unconstrained (in which we apply the ordered bijector)
+    # to constrained is monotonically increasing, i.e. order-preserving. In that case,
+    # we can form the ordered transformation as `binv ∘ OrderedBijector() ∘ b`.
+    # Similarly, if we're working with monotonically decreasing maps, we can do the same
+    # but with the addition of a sign flip before and after the ordered bijector.
+    b = bijector(d)
+    binv = inverse(b)
+    if is_monotonically_decreasing(binv)
+        ordered_b = binv ∘ SignFlip() ∘ OrderedBijector() ∘ SignFlip() ∘ b
+    elseif is_monotonically_increasing(binv)
+        ordered_b = binv ∘ OrderedBijector() ∘ b
+    else
+        throw(ArgumentError("ordered transform is currently not supported for $d."))
     end
-    return transformed(d, OrderedBijector())
+
+    return transformed(d, ordered_b)
```

Comment:
Hang on, is `transformed` even doing the right thing here? e.g. if `d` is an IID normal, then, `ordered_b` is an `OrderedBijector` that maps from unconstrained `y` to constrained `x`. So the result is an unconstrained distribution one samples from by first sampling from the IID normal and then constraining so the result is sorted. But this is not what we want. This is not the same thing as constraining `x` to be distributed according to `d` but restricted to the ordered subset of the support. Instead, what we want here is something that wraps `d` but assigns it to have the bijector `ordered_b`. Random generation is not possible, but log-density to within a normalization factor can be computed.

e.g. something like

```julia
using Distributions: Distribution, VariateForm, ValueSupport, ArrayLikeVariate

struct WithBijector{B,F<:VariateForm,S<:ValueSupport,D<:Distribution{F,S}} <: Distribution{F,S}
    dist::D
    bijector::B
end
with_bijector(d::Distribution, b) = WithBijector(d, b)
Bijectors.bijector(d::WithBijector) = d.bijector
Distributions.logpdf(d::WithBijector, x) = Distributions.logpdf(d.dist, x)
function Distributions.logpdf(d::WithBijector{<:Any, ArrayLikeVariate{N}}, x::AbstractArray{<:Real, M}) where {N, M}
    return Distributions.logpdf(d.dist, x)
end
Base.length(d::WithBijector) = length(d.dist)
```

If you then use `return with_bijector(d, inverse(ordered_b))` in this line, your example in https://github.com/TuringLang/Bijectors.jl/pull/297/files#r1597099187 seems to work.

Reply:
> This is not the same thing as constraining x to be distributed according to d but restricted to the ordered subset of the support.

Ooookay so _now_ I'm with you! I originally thought that this was just _a_ transformation that moved us to an ordered distribution, but didn't necessary have any relation to the "base" distribution (until you're previous comment on the docstring clarifying what it was supposed to do), hence never considered this to be an issue.

But yeah this is a pretty significant bug then, and has been around for a while :grimacing:

Reply:
Yeah, let's definitely get some tests in the CI that would have caught this bug.

Reply:
Indeed. Working on a fix + tests as we speak.

Reply:
Btw, something like `with_bijector` would be a convenient addition to the API for power users to manually select the bijector for a distribution. Applications include cases where the target distribution has high probability mass near a singularity of the default bijector z so one wants a custom bijector that moves the probability mass in the unconstrained space away from the singularity. Another application is programmatic benchmarking of alternative bijectors.

Reply:
> something like with_bijector would be a convenient addition to the API 

Agreed! But I think this is such a special case it warrants its own distribution (because we _can_ implement `rand` using rejection sampling for several cases used in practice). But we should also add a more general `with_bijector` method :+1: 

But I tried implementing this, and I'm still seeing fairly drastic discrepancies in the quantiles :confused: Feeling a bit out of it today though so might just have messed up the impl or tests somewhere; let me know if you see anything obvious! I pushed the changes.

Reply:
> But I tried implementing this, and I'm still seeing fairly drastic discrepancies in the quantiles 😕 Feeling a bit out of it today though so might just have messed up the impl or tests somewhere; let me know if you see anything obvious! I pushed the changes.

Ah, really? The example you posted earlier looks pretty good to me:

```julia
using Turing
using Bijectors: ordered
using LinearAlgebra
using Random: Random
using Turing.DynamicPPL: fix
using PosteriorStats
using Random

@model function demo_ordered(μ)
    k = length(μ)
    σ² ~ filldist(truncated(Normal(), lower=0), k)
    x ~ ordered(MvNormal(μ, Diagonal(σ²)))
    return (; σ², x)
end

k = 2
num_samples = 1_000_000
num_chains = 8

# Sample using NUTS.
μ = zeros(k)
σ² = ones(k)

model = fix(demo_ordered(μ); σ²)

Random.seed!(0)
chain = sample(model, NUTS(), MCMCThreads(), num_samples ÷ num_chains, num_chains)
xs_chain = permutedims(Array(chain))

# Rejection sampling.
d = MvNormal(μ, Diagonal(σ²))
xs_exact = mapreduce(hcat, 1:num_samples) do _
    while true
        xs = rand(d)
        issorted(xs) && return xs
    end
end

qts = [0.05, 0.25, 0.5, 0.75, 0.95]
qt_names = map(q -> Symbol("q$(Int(100 * q))"), qts)
stats_with_mcses = (
    Tuple(qt_names) => Base.Fix2(quantile, qts),
    (Symbol("$(qn)_mcse") => (x -> mcse(x; kind=Base.Fix2(quantile, q))) for (q, qn) in zip(qts, qt_names))...,
)
```

Comparing the quantiles, they're generally within 3 MCSEs of each other. Not really sure what the cut-off should be, but given that the MCSEs are themselves estimated using a method that uses an asymptotic approximation, I'd guess this is good enough:
```julia
julia> PosteriorStats.summarize(reshape(xs_chain', :, 1, 2), stats_with_mcses...)
SummaryStats
        q5     q25     q50      q75    q95  q5_mcse  q25_mcse  q50_mcse  q75_mcse  q95_mcse 
 1  -1.961  -1.110  -0.546  -1.e-04  0.757   0.0042    0.0022    0.0017    0.0015    0.0021
 2  -0.761  -0.001   0.544   1.107   1.955   0.0020    0.0014    0.0013    0.0015    0.0027

julia> PosteriorStats.summarize(reshape(xs_exact', :, 1, 2), stats_with_mcses...)
SummaryStats
        q5     q25     q50      q75    q95  q5_mcse  q25_mcse  q50_mcse  q75_mcse  q95_mcse 
 1  -1.955  -1.109  -0.546  -0.0003  0.763   0.0021    0.0012    0.0010    0.0011    0.0017
 2  -0.761  -0.001   0.544   1.107   1.954   0.0017    0.0011    0.0010    0.0012    0.0017
```

Reply:
My target was incorrect :man_facepalming: I implemented sequential rejection sampling, i.e. sample `x[1]` and then conditionally on this sample ordered vector using the marginals. This ofc leads to a different distribution.

Using the correct rejection sampling approach now:)

Reply:
Further progress: it now works for most things but getting correctness issues for product dist with negative support (see tests).

Reply:
Oh I just saw all your comments below :man_facepalming: